source/material_model/equation_of_state/multicomponent_incompressible.cc

```diff
@@ -36,9 +37,18 @@ namespace aspect
                const unsigned int input_index,
                MaterialModel::EquationOfStateOutputs<dim> &out) const
       {
+
+        // If adiabatic heating is used, the reference temperature used to calculate density should be the adiabatic
+        // temperature at the current position. This definition is consistent with the Extended Boussinesq Approximation.
+        const double reference_temperature = (this->include_adiabatic_heating() && this->get_adiabatic_conditions().is_initialized()
```

Comment:
Could you try to remove the second condition `is_initialized`? In the current state I am concerned that we still compute the wrong adiabatic profile (i.e. the adiabatic profile would still have a decreasing density with depth). The computation of the adiabatic profile is written in a way that even during initialization pressure and temperature at the current position are already known when evaluating the material model (you can just not ask for temperature or pressure at larger depths, or for density at the current depth). If it does not work it should trigger a signalling NaN.

Reply:
@gassmoeller - I tried removing `is_initialized?` and this gave a constant adiabatic density throughout the model domain. 

Per @anne-glerum suggestion, I added a third condition (last commit) where the `Adiabatic surface temperature` is used as the reference density prior to initialization of the adiabatic conditions. The resulting adiabatic densities are consistent with the adiabatic temperature and adiabatic surface temperature values. 

Is this a reasonable solution? 

> If it does not work it should trigger a signalling NaN.
Sorry, I didn't follow this fully. Should a signaling NaN in theory be triggered with the current code structure?